packages/sanity/src/_internal/manifest/extractWorkspaceManifest.ts

```diff
@@ -0,0 +1,502 @@
+} from 'sanity'
+
+import {
+  getCustomFields,
+  isCrossDatasetReference,
+  isCustomized,
+  isDefined,
+  isPrimitive,
+  isRecord,
+  isReference,
+  isString,
+  isType,
+} from './manifestTypeHelpers'
+import {
+  type CreateWorkspaceManifest,
+  type ManifestField,
+  type ManifestFieldset,
+  type ManifestSchemaType,
+  type ManifestSerializable,
+  type ManifestTitledValue,
+  type ManifestValidationGroup,
+  type ManifestValidationRule,
+} from './manifestTypes'
+
+interface Context {
+  schema: Schema
+}
+
+type SchemaTypeKey =
+  | keyof ArraySchemaType
+  | keyof BooleanSchemaType
+  | keyof FileSchemaType
+  | keyof NumberSchemaType
+  | keyof ObjectSchemaType
+  | keyof StringSchemaType
+  | keyof ReferenceSchemaType
+  | keyof BlockDefinition
+  | 'group' // we strip this from fields
+
+type Validation = {validation: ManifestValidationGroup[]} | Record<string, never>
+type ObjectFields = {fields: ManifestField[]} | Record<string, never>
+type SerializableProp = ManifestSerializable | ManifestSerializable[] | undefined
+type ManifestValidationFlag = ManifestValidationRule['flag']
+type ValidationRuleTransformer = (rule: RuleSpec) => ManifestValidationRule | undefined
+
+const MAX_CUSTOM_PROPERTY_DEPTH = 5
+const INLINE_TYPES = ['document', 'object', 'image', 'file']
+
+export function extractCreateWorkspaceManifest(workspace: Workspace): CreateWorkspaceManifest {
+  const serializedSchema = extractManifestSchemaTypes(workspace.schema)
+
+  return {
+    name: workspace.name,
+    title: workspace.title,
+    subtitle: workspace.subtitle,
+    basePath: workspace.basePath,
+    dataset: workspace.dataset,
+    schema: serializedSchema,
+  }
+}
+
+/**
+ * Extracts all serializable properties from userland schema types,
+ * so they best-effort can be used as definitions for Schema.compile
+. */
+export function extractManifestSchemaTypes(schema: Schema): ManifestSchemaType[] {
+  const typeNames = schema.getTypeNames()
+  const context = {schema}
+
+  const studioDefaultTypeNames = createSchema({name: 'default', types: []}).getTypeNames()
+
+  return typeNames
+    .filter((typeName) => !studioDefaultTypeNames.includes(typeName))
+    .map((typeName) => schema.get(typeName))
+    .filter((type): type is SchemaType => typeof type !== 'undefined')
+    .map((type) => transformType(type, context))
+}
+
+function transformCommonTypeFields(
+  type: SchemaType & {fieldset?: string},
+  typeName: string,
+  context: Context,
+): Omit<ManifestSchemaType, 'name' | 'title' | 'type'> {
+  const arrayProps =
+    typeName === 'array' && type.jsonType === 'array' ? transformArrayMember(type, context) : {}
+
+  const referenceProps = isReference(type) ? transformReference(type) : {}
+  const crossDatasetRefProps = isCrossDatasetReference(type)
+    ? transformCrossDatasetReference(type)
+    : {}
+
+  const objectFields: ObjectFields =
+    type.jsonType === 'object' && type.type && INLINE_TYPES.includes(typeName) && isCustomized(type)
```

Comment:
do you need `INLINE_TYPES.includes` check? i think checking for `type.jsonType === 'object'` along with your `isCustomized` check should be sufficient

Reply:
Object aliases are not allowed to re-define fields (it will throw). Checking for INLINE_TYPES is required, do determine if it is a object-with-fields definition.